### DISCUSSION TITLE
When to use Owned vs Borrowed props?

### DISCUSSION THREAD

**Discussion by dominictobias** [DioxusLabs/dioxus#1879] [Q&A] [answered]
Dioxus looks great, but this part of the docs is a bit concerning to me: I must make the decision between copying, or definitely causing render reconciliation (which could become worse over time as the render tree grows).

It would be great in this section to have some advice on when to use one or the other.

I'm also surprised that there isn't a mention of a `memo` escape hatch, even React allows us to performance tune in that way.

This is what makes Solid style observables more appealing to me: I don't need to make this difficult decision, and while Dioxus has very impressive benchmarks (higher than Leptos) and a very impressive production bundle size (again multiple times smaller), I'm concerned that it won't scale as linearly as observables would while also placing this tricky architectural burden on users.

**Comment by ealmloff** [comment, marked as answer]
> Dioxus looks great, but this part of the docs is a bit concerning to me: I must make the decision between copying, or definitely causing render reconciliation (which could become worse over time as the render tree grows).
> 
> It would be great in this section to have some advice on when to use one or the other.

The docs explain it in [this section](https://dioxuslabs.com/learn/0.4/reference/component_props#borrowed-props): *Borrowed props can be very useful, but they do not allow for memorization so they will always rerun when the parent scope is rerendered. **Because of this Borrowed Props should be reserved for components that are cheap to rerun or places where cloning data is an issue. Using Borrowed Props everywhere will result in large parts of your app rerunning every interaction.***

> I'm also surprised that there isn't a mention of a memo escape hatch, even React allows us to performance tune in that way.

In dioxus, components are memo by default. Borrowed props opt out of memo. If you want to opt out of memo for non-borrowed props, you can implement partialeq as always false for your props struct

> This is what makes Solid style observables more appealing to me: I don't need to make this difficult decision, and while Dioxus has very impressive benchmarks (higher than Leptos) and a very impressive production bundle size (again multiple times smaller), I'm concerned that it won't scale as linearly as observables would while also placing this tricky architectural burden on users.

As far as I know in leptos and the newest version of sycamore, you always need to clone or move props into children. You don't have the option of borrowed props.

That said, we have been exploring different state management systems in dioxus. [Dioxus-signals](https://crates.io/crates/dioxus-signals) was released in 0.4.3. It provides `Copy` state with local subscriptions. https://github.com/DioxusLabs/dioxus/pull/1791 makes signals the default, removes borrowed props entirely, makes `ScopeState` global, and removes all `unsafe` blocks from core.

**Reply by dominictobias** [reply]
Thank you that’s really helpful 🙏 Especially with this clarification and move to signals Dioxus seems like the finest combination of all the things that’s out there, the framework I've been searching for my whole life :>
